extprocconfig/extprocconfig.go

```diff
@@ -0,0 +1,121 @@
+// This is a public package so that the external processor can be testable without
+// depending on the Envoy Gateway as well as it can be used outside the Envoy AI Gateway.
+//
+// This configuration must be decoupled from the Envoy Gateway types as well as its implementation
+// details.
+package extprocconfig
+
+import (
+	"os"
+
+	"k8s.io/apimachinery/pkg/util/yaml"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+// Config is the configuration for the external processor.
+//
+// The configuration is loaded from a file path specified via the command line flag -configPath to the external processor.
+//
+// # Example configuration:
+//
+//	inputSchema:
+//	  schema: OpenAI
+//	backendRoutingHeaderKey: x-backend-name
+//	modelNameHeaderKey: x-model-name
+//	rules:
+//	- backends:
+//	  - name: kserve
+//	    weight: 1
+//	    outputSchema:
+//	      schema: OpenAI
+//	  - name: awsbedrock
+//	    weight: 10
+//	    outputSchema:
+//	      schema: AWSBedrock
+//	  headers:
+//	  - name: x-model-name
+//	    value: llama3.3333
+//	- backends:
+//	  - name: openai
+//	    outputSchema:
+//	      schema: OpenAI
+//	  headers:
+//	  - name: x-model-name
+//	    value: gpt4.4444
+//
+// where the input of the external processor is in the OpenAI schema, the model name is populated in the header x-model-name,
+// The model name header `x-model-name` is used in the header matching to make the routing decision. **After** the routing decision is made,
+// the selected backend name is populated in the header `x-backend-name`. For example, when the model name is `llama3.3333`,
+// the request is routed to either backends `kserve` or `awsbedrock` with weights 1 and 10 respectively, and the selected
+// backend, say `awsbedrock`, is populated in the header `x-backend-name`.
+//
+// From Envoy configuration perspective, configuring the header matching based on `x-backend-name` is enough to route the request to the selected backend.
+// That is because the matching decision is made by the external processor and the selected backend is populated in the header `x-backend-name`.
+type Config struct {
+	// InputSchema specifies the API schema of the input format of requests to the external processor.
+	InputSchema VersionedAPISchema `yaml:"inputSchema"`
+	// ModelNameHeaderKey is the header key to be populated with the model name by the external processor.
+	ModelNameHeaderKey string `yaml:"modelNameHeaderKey"`
+	// BackendRoutingHeaderKey is the header key to be populated with the backend name by the external processor
+	// **after** the routing decision is made by the external processor using Rules.
+	BackendRoutingHeaderKey string `yaml:"backendRoutingHeaderKey"`
+	// Rules is the routing rules to be used by the external processor to make the routing decision.
+	// Inside the routing rules, the header ModelNameHeaderKey may be used to make the routing decision.
+	Rules []RouteRule `yaml:"rules"`
```

Comment:
Why the rules are configured in the extproc configmap ?

Reply:
these rules are expect to be updated quite frequently as users add or remove models endpoints, or we are translating the LLMroute to the config for extproc ?

Reply:
> or we are translating the LLMroute to the config for extproc ?

yes, this is because of the requirement about not restarting the extproc on the LLMRoute/etc updates